controllers/component_customizer_test.go

```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	configv1alpha1 "k8s.io/component-base/config/v1alpha1"
+	"k8s.io/utils/pointer"
+	operatorv1 "sigs.k8s.io/cluster-api-operator/api/v1alpha1"
+	configv1 "sigs.k8s.io/controller-runtime/pkg/config/v1alpha1"
+)
+
+func intPtr(v int) *int {
+	return &v
+}
+
+func TestCustomizeDeployment(t *testing.T) {
+	sevenHours, _ := time.ParseDuration("7h")
+	memTestQuantity, _ := resource.ParseQuantity("16Gi")
+	managerDepl := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "manager",
+			Namespace: metav1.NamespaceSystem,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "manager",
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{
+						Name:  "manager",
+						Image: "k8s.gcr.io/a-manager:1.6.2",
+						Env: []corev1.EnvVar{
+							{
+								Name:  "test1",
+								Value: "value1",
+							},
+						},
+						Args: []string{"--webhook-port=2345"},
+						LivenessProbe: &corev1.Probe{
+							ProbeHandler: corev1.ProbeHandler{
+								HTTPGet: &corev1.HTTPGetAction{
+									Path: "/healthz",
+									Port: intstr.FromString("healthz"),
+								},
+							},
+						},
+						ReadinessProbe: &corev1.Probe{
+							ProbeHandler: corev1.ProbeHandler{
+								HTTPGet: &corev1.HTTPGetAction{
+									Path: "/readyz",
+									Port: intstr.FromString("healthz"),
+								},
+							},
+						},
+					}},
+				},
+			},
+		},
+	}
+	tests := []struct {
+		name  string
+		dSpec *operatorv1.DeploymentSpec
+		mSpec *operatorv1.ManagerSpec
+		dIn   *appsv1.Deployment
+		dOut  *appsv1.Deployment
+	}{
+		{
+			name:  "empty",
+			dSpec: &operatorv1.DeploymentSpec{},
+			dIn:   managerDepl.DeepCopy(),
+			dOut:  managerDepl,
+		},
+		{
+			name: "all deployment options",
+			dSpec: &operatorv1.DeploymentSpec{
```

Comment:
It's hard to understand what changes with each of these test cases right now, having some helpers that take in a deployment and modify a specific field would make this easier to understand what we are trying to do with each test